optimade/filtertransformers/mongo.py

```diff
@@ -81,20 +77,52 @@ def expression_phrase(self, arg):
         # expression_phrase: [ NOT ] ( comparison | "(" expression ")" )
         return self._recursive_expression_phrase(arg)
 
-    def property_first_comparison(self, arg):
+    @v_args(inline=True)
+    def property_first_comparison(self, quantity, query):
         # property_first_comparison: property ( value_op_rhs | known_op_rhs | fuzzy_string_op_rhs | set_op_rhs |
         # set_zip_op_rhs | length_op_rhs )
 
         # Awkwardly, MongoDB will match null fields in $ne filters,
         # so we need to add a check for null equality in evey $ne query.
-        if "$ne" in arg[1]:
-            return {"$and": [{arg[0]: arg[1]}, {arg[0]: {"$ne": None}}]}
-
-        return {arg[0]: arg[1]}
+        if "$ne" in query:
+            return {"$and": [{quantity: query}, {quantity: {"$ne": None}}]}
+
+        # Check if a $size query is being made (indicating a length_op_rhs filter); if so, check for
+        # a defined length alias to replace the $size call with the corresponding filter on the
+        # length quantity then carefully merge the two queries.
+        #
+        # e.g. `("elements", {"$size": 2, "$all": ["Ag", "Au"]})` should become
+        # `{"elements": {"$all": ["Ag", "Au"]}, "nelements": 2}` if the `elements` -> `nelements`
+        # length alias is defined.
+        if "$size" in query:
+            if (
+                getattr(self.backend_mapping.get(quantity), "length_quantity", None)
+                is not None
+            ):
+                size_query = {
+                    self.backend_mapping[
+                        quantity
+                    ].length_quantity.backend_field: query.pop("$size")
+                }
+
+                final_query = {}
+                if query:
+                    final_query = {quantity: query}
+                for q in size_query:
+                    if q in query:
+                        query[q].update(size_query[q])
```

Comment:
```suggestion
                    if q in final_query:
                        final_query[q].update(size_query[q])
```

(from https://github.com/Materials-Consortia/optimade-python-tools/pull/797/files/2776057a7a1fac38f6cd51838df03423ac00eea5#r642375176)